unumpy/_multimethods.py

```diff
@@ -816,25 +796,442 @@ def nancumsum(a, axis=None, dtype=None, out=None):
+def _self_weights_argreplacer(args, kwargs, dispatchables):
+    def replacer(*args, weights=None, **kwargs):
+        return dispatchables[0:-1], dict(weights=dispatchables[-1], **kwargs)
+
+    return replacer(*args, **kwargs)
+
+
+def _average_default(a, axis=None, weights=None, returned=False):
+    if weights is None:
+        weights = ones(a.shape)
+
+    if a.shape != weights.shape:
+        if axis is None:
+            raise TypeError(
+                "Axis must be specified when shapes of a and weights differ."
+            )
+        if ndim(weights) != 1:
+            raise TypeError("1D weights expected when shapes of a and weights differ.")
+        if weights.shape[0] != a.shape[axis]:
+            raise ValueError("Length of weights not compatible with specified axis.")
+
+        weights = broadcast_to(weights, (1,) * (ndim(a) - 1) + weights.shape)
+        weights = swapaxes(weights, -1, axis)
+        weights = weights * ones(a.shape)
+
+    a, _ = _ureduce(a * weights, axis)
+
+    sum_of_weights = sum(weights, axis=axis)
+
+    a = sum(a, axis=-1) / sum_of_weights
+
+    if returned:
+        return a, sum_of_weights
+    else:
+        return a
+
+
+@create_numpy(_self_weights_argreplacer, default=_average_default)
+@all_of_type(ndarray)
+def average(a, axis=None, weights=None, returned=False):
+    return (a, weights)
+
+
+def _mean_default(a, axis=None, dtype=None, out=None, keepdims=False):
+    a, dims = _ureduce(a, axis)
+
+    if dtype is None:
+        if a.dtype.type == "i":
+            dtype = float
+        else:
+            dtype = a.dtype
+
+    N = a.shape[-1]
+
+    a = sum(a, axis=-1, dtype=dtype) / N
+
+    if keepdims:
+        a = a.reshape(dims)
+
+    if out is None:
+        return a
+
+    if a.shape != out.shape:
+        raise ValueError("out parameter must have the same shape as the output")
+
+    copyto(out, a, casting="unsafe")
+    return out
+
+
+@create_numpy(_self_dtype_argreplacer, default=_mean_default)
+@all_of_type(ndarray)
+def mean(a, axis=None, dtype=None, out=None, keepdims=False):
+    return (a, mark_dtype(dtype), mark_non_coercible(out))
+
+
+@create_numpy(
+    _self_dtype_argreplacer,
+    default=lambda a, axis=None, dtype=None, out=None, ddof=0, keepdims=False: sqrt(
+        var(a, axis=axis, dtype=dtype, out=out, ddof=ddof, keepdims=keepdims)  # type: ignore
+    ),
+)
+@all_of_type(ndarray)
+def std(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
+    return (a, mark_dtype(dtype), mark_non_coercible(out))
+
+
+def _var_default(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
+    if ddof != 0:
+        return NotImplemented
+
+    a, dims = _ureduce(a, axis)
+
+    if dtype is None:
+        if a.dtype.type == "i":
+            dtype = float
+        else:
+            dtype = a.dtype
+
+    N = a.shape[-1] - ddof
+
+    x = sum(a ** 2, axis=-1, dtype=dtype) / N
+    y = sum(a, axis=-1, dtype=dtype) / N
+
+    a = x - y ** 2
```

Comment:
```suggestion
    x = sum(a ** 2, axis=-1, dtype=dtype)
    y = sum(a, axis=-1, dtype=dtype)

    a = (x - y ** 2) / N
```

Reply:
This doesn't fix it. 😕

Reply:
Can you raise an issue on the NumPy issue tracker saying it gives weird results?

Reply:
Or look at the NumPy source code?

Reply:
> Can you raise an issue on the NumPy issue tracker saying it gives weird results?

Yes, I can do that. In that case I should probably compare results. Is our implementation giving the right ones?

> Or look at the NumPy source code?

It does different computations. I don't fully understand it but I will take a closer look.

Reply:
I think you'll find it's equivalent to this:
```python
x = mean(a**2, axis=axis, dtype=dtype)
y = mean(a, axis=axis, dtype=dtype)
a = (x - y**2) * (a.shape[-1] / N)
```

Reply:
That would be equivalent to the square of the [standard deviation](https://en.wikipedia.org/wiki/Standard_deviation) formula.

Reply:
@peterbell10 Thanks for the insight!

Reply:
I've pushed the changes that fix this.